examples/drivers/windows/otLwf/thread.c

```diff
@@ -253,9 +306,124 @@ void otPlatFree(void *aPtr)
+    for (uint16_t i = 1; i < pFilter->otBufferPoolBufferCount; i++)
+    {
+        struct BufferHeader* curBuf =
+            (struct BufferHeader*)&bufPool->Buffers[i * pFilter->otBufferSize];
+
+        prevBuf->mNext = curBuf;
+        prevBuf = curBuf;
+    }
+
+    LogVerbose(DRIVER_DEFAULT, "Allocated new buffer pool (%d bytes)!", pFilter->otBufferPoolByteSize);
+
+    return bufPool;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+struct BufferHeader* GetNextFreeBufferFromPool(_In_ PMS_FILTER pFilter)
+{
+    // Immediately return if we have hit our limit
+    if (pFilter->otBuffersLeft == 0) return NULL;
+
+    // If we don't have any free buffers left, allocate another pool
+    if (pFilter->otFreeBuffers == NULL)
+    {
+        BufferPool *newPool = AllocBufferPool(pFilter);
+        if (newPool == NULL) return NULL; // Out of physical memory
+
+        // Push on top of the pool list
+        newPool->Next = pFilter->otBufferPoolHead;
+        pFilter->otBufferPoolHead = newPool;
+
+        // Set the free buffer list
+        pFilter->otFreeBuffers = (struct BufferHeader*)newPool->Buffers;
+    }
+
+    // Pop the top free buffer
+    struct BufferHeader* buffer = pFilter->otFreeBuffers;
+    pFilter->otFreeBuffers = pFilter->otFreeBuffers->mNext;
+    pFilter->otBuffersLeft--;
+    buffer->mNext = NULL;
+    return buffer;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void otPlatMessagePoolInit(_In_ otInstance *otCtx, uint16_t aMinNumFreeBuffers, size_t aBufferSize)
+{
+    NT_ASSERT(otCtx);
+    PMS_FILTER pFilter = otCtxToFilter(otCtx);
+
+    LogFuncEntry(DRIVER_DEFAULT);
+    UNREFERENCED_PARAMETER(aMinNumFreeBuffers);
+
+    // Initialize parameters
+    pFilter->otBufferSize = (uint16_t)aBufferSize;
+    pFilter->otBufferPoolByteSize = (uint16_t)(kPageSize * kPagesPerBufferPool);
+    pFilter->otBufferPoolBufferCount = (uint16_t)((pFilter->otBufferPoolByteSize - sizeof(BufferPool)) / aBufferSize);
+    pFilter->otBuffersLeft = kMaxPagesForBufferPools * pFilter->otBufferPoolBufferCount;
+
+    // Allocate first pool
+    pFilter->otBufferPoolHead = AllocBufferPool(pFilter);
+    ASSERT(pFilter->otBufferPoolHead); // Should this API allow for failure ???
+
+    // Set initial free buffer list
+    pFilter->otFreeBuffers = (struct BufferHeader*)pFilter->otBufferPoolHead->Buffers;
+
+    LogFuncExit(DRIVER_DEFAULT);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+struct BufferHeader *otPlatMessagePoolNew(_In_ otInstance *otCtx)
+{
+    NT_ASSERT(otCtx);
+    PMS_FILTER pFilter = otCtxToFilter(otCtx);
+    return GetNextFreeBufferFromPool(pFilter);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void otPlatMessagePoolFree(_In_ otInstance *otCtx, _In_ struct BufferHeader *aBuffer)
+{
+    NT_ASSERT(otCtx);
+    PMS_FILTER pFilter = otCtxToFilter(otCtx);
+
+    // Put buffer back on the list
+    aBuffer->mNext = pFilter->otFreeBuffers;
+    pFilter->otFreeBuffers = aBuffer;
+    pFilter->otBuffersLeft++;
```

Comment:
The current design doesn't free `BufferPool` objects until the entire interface is removed. I'm not sure it's really needed, nor do I have a good algorithm to achieve it.